qualtran/bloqs/data_loading/qrom_abc.py

```diff
@@ -0,0 +1,294 @@
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+"""Base class for Bloqs implementing a QROM (Quantum read-only memory) circuit."""
+import abc
+from functools import cached_property
+from typing import Dict, Tuple, Type, TypeVar, Union
+
+import attrs
+import cirq
+import numpy as np
+from numpy.typing import ArrayLike, NDArray
+
+from qualtran import BoundedQUInt, QAny, Register
+from qualtran.simulation.classical_sim import ClassicalValT
+from qualtran.symbolics import bit_length, is_symbolic, shape, Shaped, SymbolicInt
+
+QROMT = TypeVar('QROMT', bound='QROMABC')
+
+
+@cirq.value_equality(distinct_child_types=True)
+@attrs.frozen
+class QROMABC(abc.ABC):
+    r"""Bloq to load `data[l]` in the target register when the selection stores an index `l`.
+
+    ## Overview
+    The action of a QROM can be described as
+    $$
+            \text{QROM}_{s_1, s_2, \dots, s_K}^{d_1, d_2, \dots, d_L}
+            |s_1\rangle |s_2\rangle \dots |s_K\rangle
+            |0\rangle^{\otimes b_1} |0\rangle^{\otimes b_2} \dots |0\rangle^{\otimes b_L}
+            \rightarrow
+            |s_1\rangle |s_2\rangle \dots |s_K\rangle
+            |d_1[s_1, s_2, \dots, s_k]\rangle
+            |d_2[s_1, s_2, \dots, s_k]\rangle \dots
+            |d_L[s_1, s_2, \dots, s_k]\rangle
+    $$
+
+    There two high level parameters that control the behavior of a QROM are -
+
+    1. Shape of the classical dataset to be loaded ($\text{data.shape} = (S_1, S_2, ..., S_K)$).
+    2. Number of distinct datasets to be loaded ($\text{data.bitsizes} = (b_1, b_2, ..., b_L)$).
+
+    Each of these have an effect on the cost of the QROM. The `data_or_shape` parameter stores
+    either
+    1. A numpy array of shape $(L, S_1, S_2, ..., S_K)$ when $L$ classical datasets, each of
+       shape $(S_1, S_2, ..., S_K)$ and bitsizes $(b_1, b_2, ..., b_L)$ are to be loaded and
+       the classical data is available to instantiate the QROM bloq. In this case, the helper
+       builder `QROM.build_from_data(data_1, data_2, ..., data_L)` can be used to build the QROM.
+
+    2. A `Shaped` object that stores a (potentially symbolic) tuple $(L, S_1, S_2, ..., S_K)$
+       that represents the number of classical datasets `L=data_or_shape.shape[0]` and
+       their shape `data_shape=data_or_shape.shape[1:]` to be loaded by this QROM. This is used
+       to instantiate QROM bloqs for symbolic cost analysis where the exact data to be loaded
+       is not known. In this case, the helper builder `QROM.build_from_bitsize` can be used
+       to build the QROM.
+
+    ### Shape of the classical dataset to be loaded.
+    QROM bloq supports loading multidimensional classical datasets. In order to load a data
```

Comment:
Not The or A

Reply:
Can you clarify?

Reply:
I'm not sure anymore lol. I think it's was meant to be nit: The (or A) QROM bloq supports...